lib/puppet/util/character_encoding.rb

```diff
@@ -3,78 +3,84 @@
-          # encoding, fail. It requires user remediation.
-          raise EncodingError
+          # If the string comes to us as BINARY encoded, we don't know what it
+          # started as. However, to encode! we need a starting place, and our
+          # best guess is whatever the system currently is (default_external).
+          # So set external_encoding to default_external before we try to
+          # transcode to UTF-8.
+          string.force_encoding(Encoding.default_external) if original_encoding == Encoding::BINARY
+          string.encode!(Encoding::UTF_8)
         end
       rescue EncodingError => detail
+        # Ensure the string retains it original external encoding since
+        # we've failed
+        string.force_encoding(original_encoding) if original_encoding == Encoding::BINARY
         # Catch both our own self-determined failure to transcode as well as any
         # error on ruby's part, ie Encoding::UndefinedConversionError on a
         # failure to encode!.
-        Puppet.debug(_("%{error}: %{value} is not valid UTF-8 and cannot be transcoded by Puppet.") %
+        Puppet.debug(_("%{error}: %{value} cannot be transcoded by Puppet.") %
           { error: detail.inspect, value: string.dump })
-        return nil
+      end
+      return nil
+    end
+
+    # Given a string, tests if that string's bytes represent valid UTF-8, and
+    # if so return a copy of the string with external enocding set to UTF-8.
+    # Does not modify the byte representation of the string. If the string does
+    # not represent valid UTF-8, does not set the external encoding.
+    #
+    # This method is intended for situations where we do not believe that the
+    # encoding associated with a string is an accurate reflection of its actual
+    # bytes, i.e., effectively when we believe Ruby is incorrect in its assertion
+    # of the encoding of the string.
+    #
+    # @api public
+    # @param [String] string to set external encoding (re-label) to utf-8
+    # @return [String] a copy of string with external encoding set to utf-8
+    def override_encoding_to_utf_8(string)
+      if valid_utf_8_bytes?(string)
+        string.dup.force_encoding(Encoding::UTF_8)
+      else
+        Puppet.debug(_("%{value} is not valid UTF-8 and result of overriding encoding would be invalid.") % { value: string.dump })
+        string
+      end
+    end
+
+    # Given a string, return a copy of that string with any invalid byte
+    # sequences in its current encoding replaced with "?"
+    # @param string a string to remove invalid byte sequences from
+    # @return a copy of string invalid byte sequences replaced by "?" character
+    # @note does not modify encoding, but new string will have different bytes
+    #   from original
+    def scrub(string)
+      if String.method_defined?(:scrub)
+        string.scrub
+      else
+        string.chars.map { |c| c.valid_encoding? ? c : "?"}.join
```

Comment:
scrub isn't present in 1.9.3, and various iterations on encode with replacement characters failed to work. supplying utf-8 for the first two args to encode is a no-op, and supplying binary as the second arg causes any non-ascii character to be replaced regardless of the original encoding or validity of the string...